---
title: "Pathway Analysis"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Preparing for DESeq2
About our Input Data

The data for for hands-on session comes from GEO entry: GSE37704, which is associated with the following publication:

Trapnell C, Hendrickson DG, Sauvageau M, Goff L et al. "Differential analysis of gene regulation at transcript resolution with RNA-seq". Nat Biotechnol 2013 Jan;31(1):46-53. PMID: 23222703

The authors report on differential analysis of lung fibroblasts in response to loss of the developmental transcription factor HOXA1. Their results and others indicate that HOXA1 is required for lung fibroblast and HeLa cell cycle progression. In particular their analysis show that "loss of HOXA1 results in significant expression level changes in thousands of individual transcripts, along with isoform switching events in key regulators of the cell cycle". For our session we have used their Sailfish gene-level estimated counts and hence are restricted to protein-coding genes only.
```{r}
library(DESeq2)
```

```{r}
meta <- "GSE37704_metadata.csv"
counts <- "GSE37704_featurecounts.csv"

colData <- read.csv(meta)
head(colData)
```

```{r}
countData <- read.csv(counts, row.names = 1)
head(countData)
```

Remove length column from countData
```{r}
countData <- countData[,-1]
head(countData)
```
```{r}
nrow(countData)
```

Remove rows with zeros for all samples
```{r}
countData <- countData[rowSums(countData)>0,]
nrow(countData)
```

## PCA exploration

```{r}
pc <- prcomp(t(countData))
summary(pc)

```

```{r}
plot(pc$x[,1], pc$x[,2], col = colData$condition, pch = 16)
```

```{r}
library(ggplot2)

theme(panel.background = element_rect(fill = "transparent"), # bg of the panel
       plot.background = element_rect(fill = "transparent", color = NA), # bg of the plot
       axis.title.x = element_text(size=14, face="bold"),
       axis.title.y = element_text(size=14, face="bold"),
       axis.text.x = element_text(face="bold", size=14),
       axis.text.y = element_text(face="bold", size=14),
       panel.grid.major = element_blank(), # get rid of major grid
       panel.grid.minor = element_blank(), # get rid of minor grid
       legend.background = element_rect(fill = "transparent"), # get rid of legend bg
       legend.box.background = element_rect(fill = "transparent"), # get rid of legend panel bg
       axis.line = element_line(color="black"))
options(scipen = 500)
forgg <- as.data.frame(pc$x)
ggplot(forgg, aes(x = forgg$PC1, y = forgg$PC2, color = colData$condition)) +
  geom_point() +
  xlab("PC1") +
  ylab("PC2") 
  

```
## DESeq2

```{r}
dds = DESeqDataSetFromMatrix(countData=countData,
                             colData=colData,
                             design=~condition)
dds = DESeq(dds)

```

```{r}
res <- results(dds)
res
```

## Volcano Plot Summary
```{r}
mycols <- rep("gray", nrow(res))
mycols[ abs(res$log2FoldChange) > 2 ]  <- "red" 

inds <- (res$padj < 0.01) & (abs(res$log2FoldChange) > 2 )
mycols[ inds ] <- "blue"

# Volcano plot with custom colors 
plot( res$log2FoldChange,  -log(res$padj), 
 col=mycols, ylab="-Log(P-value)", xlab="Log2(FoldChange)", pch = 1 )

# Cut-off lines
abline(v=c(-2,2), col="black", lty=2)
abline(h=-log(0.1), col="black", lty=2)
```

## Add annotation

```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")

columns(org.Hs.eg.db)

res$symbols <- mapIds(org.Hs.eg.db, 
                      keys = row.names(res),
                      keytype = "ENSEMBL",
                      column = "SYMBOL",
                      multiVals="first")

res$entrez <- mapIds(org.Hs.eg.db, 
                      keys = row.names(res),
                      keytype = "ENSEMBL",
                      column = "ENTREZID",
                      multiVals="first")

res$name <- mapIds(org.Hs.eg.db, 
                      keys = row.names(res),
                      keytype = "ENSEMBL",
                      column = "GENENAME",
                      multiVals="first")

```
```{r}

head(res, 10)
```

Reorder by padj and write results to a new .csv
```{r}
res <- res[order(res$pvalue),]
write.csv(res, file="deseq_results.csv")

```

## Pathway Analysis

```{r}
library(pathview)
```

```{r}
library(gage)
library(gageData)

data(kegg.sets.hs)
data(sigmet.idx.hs)

# Focus on signaling and metabolic pathways only
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]

# Examine the first 3 pathways
head(kegg.sets.hs, 3)
```

The main gage() function requires a named vector of fold changes, where the names of the values are the Entrez gene IDs.

Note that we used the mapIDs() function above to obtain Entrez gene IDs (stored in res\$entrez) and we have the fold change results from DESeq2 analysis (stored in res\$log2FoldChange).

```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges, 10)
```

Now, letâ€™s run the gage pathway analysis.
```{r}
# Get the results
keggres = gage(foldchanges, gsets=kegg.sets.hs)

attributes(keggres)
head(keggres$less)
```

Visualize with pathview
```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04110")
```

![](./hsa04110.pathview.png)


A slightly different representation, in PDF format
```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04110", kegg.native=FALSE)
```

Now, let's process our results a bit more to automagicaly pull out the top 5 upregulated pathways, then further process that just to get the pathway IDs needed by the pathview() function. We'll use these KEGG pathway IDs for pathview plotting below.

```{r}
## Focus on top 5 upregulated pathways here for demo purposes only
keggrespathways <- rownames(keggres$greater)[1:5]

# Extract the 8 character long IDs part of each string
keggresids = substr(keggrespathways, start=1, stop=8)
keggresids
```

Finally, lets pass these IDs in keggresids to the pathview() function to draw plots for all the top 5 pathways.
```{r}
pathview(gene.data=foldchanges, pathway.id=keggresids, species="hsa")
```